vpr/src/device/rr_graph_builder.h

```diff
@@ -40,6 +38,63 @@ class RRGraphBuilder {
     t_rr_graph_storage& node_storage();
     /** @brief Return a writable object for update the fast look-up of rr_node */
     RRSpatialLookup& node_lookup();
+    /** @brief Return a writable object for rr_node_metadata */
+    MetadataStorage<int>& rr_node_metadata();
+    /** @brief Return a writable object for rr_edge_metadata */
+    MetadataStorage<std::tuple<int, int, short>>& rr_edge_metadata();
+
+    MetadataStorage<int> rr_node_metadata_data() const;
+
+    MetadataStorage<std::tuple<int, int, short>> rr_edge_metadata_data() const;
+    //rr_node_metadata_clear
+    inline void rr_node_metadata_clear() {
+        rr_node_metadata_.clear();
+    }
+    inline void rr_edge_metadata_clear() {
+        rr_edge_metadata_.clear();
+    }
+    // rr_edge_metadata_size
+    inline size_t rr_edge_metadata_size() const {
+        return rr_edge_metadata_.size();
+    }
+    inline size_t rr_node_metadata_size() const {
+        return rr_node_metadata_.size();
+    }
+
+    //rr_node_metadata_find
```

Comment:
Templates should not be placed with regular APIs. My suggestion is to move to the head of a class definition.
See example:

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/36789e456ab292038b8518bf2c835c9b9dcaaa0e/vpr/src/device/rr_graph_obj.h#L213-L224

Reply:
If I make the template as a class template over the RR_graph_builder.h as it is on the metadata_storage.h. Almost all the API's in the class become unknown to the locations where they have been used. Should I make the changes in all the API's or I am doing some thing wrong as I am not worked much with the templates. 
https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/36789e456ab292038b8518bf2c835c9b9dcaaa0e/vpr/src/base/metadata_storage.h#L18-L19
Please guide me how to resolve this issue.